File: backend/app/routers/ai.py
from fastapi import APIRouter, Body, HTTPException, Depends
from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field
from ..services.ai import ai_service
from ..db import get_db_connection
from ..auth import get_current_user, User, require_auth

router = APIRouter()

class PromptModel(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    system_prompt: str = Field(..., min_length=1, max_length=10000)
    user_prompt: str = Field(..., min_length=1, max_length=10000)
    is_default: bool = False

@router.post("/ai/analyze_fund")
async def analyze_fund(
    fund_info: Dict[str, Any] = Body(...),
    prompt_id: int = Body(None),
    current_user: User = Depends(require_auth)
):
    """
    分析基金（需要认证）
    """
    return await ai_service.analyze_fund(fund_info, prompt_id=prompt_id, user_id=current_user.id)

@router.get("/ai/prompts")
def get_prompts(current_user: User = Depends(require_auth)):
    """
    获取 AI 提示词模板（按用户隔离）
    """
    conn = get_db_connection()
    cursor = conn.cursor()

    cursor.execute("""
        SELECT id, name, system_prompt, user_prompt, is_default, created_at, updated_at
        FROM ai_prompts
        WHERE user_id = ?
        ORDER BY is_default DESC, id ASC
    """, (current_user.id,))

    prompts = [dict(row) for row in cursor.fetchall()]
    return {"prompts": prompts}

@router.post("/ai/prompts")
def create_prompt(
    data: PromptModel,
    current_user: User = Depends(require_auth)
):
    """
    创建新的 AI 提示词模板（需要认证）
    """
    conn = get_db_connection()
    cursor = conn.cursor()

    # If this is set as default, unset other defaults for this user
    if data.is_default:
        cursor.execute("UPDATE ai_prompts SET is_default = 0 WHERE user_id = ?", (current_user.id,))

    # Insert with user_id
    cursor.execute("""
        INSERT INTO ai_prompts (name, system_prompt, user_prompt, user_id, is_default)
        VALUES (?, ?, ?, ?, ?)
    """, (data.name, data.system_prompt, data.user_prompt, current_user.id, 1 if data.is_default else 0))

    prompt_id = cursor.lastrowid
    conn.commit()

    return {"ok": True, "id": prompt_id}

@router.put("/ai/prompts/{prompt_id}")
def update_prompt(
    prompt_id: int,
    data: PromptModel,
    current_user: User = Depends(require_auth)
):
    """
    更新 AI 提示词模板（需要认证，只能更新自己的）
    """
    conn = get_db_connection()
    cursor = conn.cursor()

    # Check if prompt exists and belongs to user
    cursor.execute("SELECT id FROM ai_prompts WHERE id = ? AND user_id = ?", (prompt_id, current_user.id))

    if not cursor.fetchone():
        raise HTTPException(status_code=404, detail="Prompt not found or access denied")

    # If this is set as default, unset other defaults for this user
    if data.is_default:
        cursor.execute("UPDATE ai_prompts SET is_default = 0 WHERE id != ? AND user_id = ?", (prompt_id, current_user.id))

    cursor.execute("""
        UPDATE ai_prompts
        SET name = ?, system_prompt = ?, user_prompt = ?, is_default = ?, updated_at = CURRENT_TIMESTAMP
        WHERE id = ?
    """, (data.name, data.system_prompt, data.user_prompt, 1 if data.is_default else 0, prompt_id))

    conn.commit()

    return {"ok": True}

@router.delete("/ai/prompts/{prompt_id}")
def delete_prompt(
    prompt_id: int,
    current_user: User = Depends(require_auth)
):
    """
    删除 AI 提示词模板（需要认证，只能删除自己的）
    """
    conn = get_db_connection()
    cursor = conn.cursor()

    # Check if prompt exists and belongs to user
    cursor.execute("SELECT is_default FROM ai_prompts WHERE id = ? AND user_id = ?", (prompt_id, current_user.id))

    row = cursor.fetchone()

    if not row:
        raise HTTPException(status_code=404, detail="Prompt not found or access denied")

    if row["is_default"]:
        raise HTTPException(status_code=400, detail="不能删除默认模板")

    cursor.execute("DELETE FROM ai_prompts WHERE id = ?", (prompt_id,))
    conn.commit()

    return {"ok": True}

@router.get("/ai/analysis_history")
def get_analysis_history(
    fund_code: str,
    account_id: int = 1,
    limit: int = 20,
    offset: int = 0,
    prompt_id: Optional[int] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    current_user: User = Depends(require_auth)
):
    """
    获取 AI 分析历史记录列表（不返回 markdown，减少数据量）
    """
    conn = get_db_connection()
    cursor = conn.cursor()

    # Build query
    query = """
        SELECT id, fund_code, fund_name, prompt_name, status, created_at
        FROM ai_analysis_history
        WHERE account_id = ? AND fund_code = ? AND user_id = ?
    """
    params = [account_id, fund_code, current_user.id]

    # Add optional filters
    if prompt_id is not None:
        query += " AND prompt_id = ?"
        params.append(prompt_id)

    if date_from:
        query += " AND created_at >= ?"
        params.append(date_from)

    if date_to:
        query += " AND created_at <= ?"
        params.append(date_to)

    query += " ORDER BY created_at DESC LIMIT ? OFFSET ?"
    params.extend([limit, offset])

    cursor.execute(query, params)
    records = [dict(row) for row in cursor.fetchall()]

    return {"records": records}

@router.get("/ai/analysis_history/{history_id}")
def get_analysis_history_detail(
    history_id: int,
    current_user: User = Depends(require_auth)
):
    """
    获取单条 AI 分析历史记录详情（包含 markdown）
    """
    conn = get_db_connection()
    cursor = conn.cursor()

    cursor.execute("""
        SELECT * FROM ai_analysis_history
        WHERE id = ? AND user_id = ?
    """, (history_id, current_user.id))

    row = cursor.fetchone()

    if not row:
        raise HTTPException(status_code=404, detail="History record not found or access denied")

    return dict(row)

@router.delete("/ai/analysis_history/{history_id}")
def delete_analysis_history(
    history_id: int,
    current_user: User = Depends(require_auth)
):
    """
    删除 AI 分析历史记录（只能删除自己的）
    """
    conn = get_db_connection()
    cursor = conn.cursor()

    # Check if record exists and belongs to user
    cursor.execute("SELECT id FROM ai_analysis_history WHERE id = ? AND user_id = ?", (history_id, current_user.id))

    if not cursor.fetchone():
        raise HTTPException(status_code=404, detail="History record not found or access denied")

    try:
        cursor.execute("DELETE FROM ai_analysis_history WHERE id = ?", (history_id,))
        conn.commit()
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail="Failed to delete record")

    return {"ok": True}

File: worker/src/routes/ai.js
/**
 * AI API routes (Hono)
 * Migrated from backend/app/routers/ai.py
 */

import { Hono } from 'hono';
import { getCurrentUser } from '../auth.js';
import { aiService } from '../services/ai.js';

const ai = new Hono();

// POST /api/ai/analyze
ai.post('/analyze', async (c) => {
  const user = await getCurrentUser(c);
  if (!user) return c.json({ detail: '未登录' }, 401);

  const body = await c.req.json();
  const { fund_id, fund_name, prompt_id, account_id } = body;

  if (!fund_id) return c.json({ detail: '需要指定基金代码' }, 400);

  try {
    const result = await aiService.analyzeFund(c.env.DB, c.env, {
      id: fund_id,
      name: fund_name,
      account_id: account_id || 1,
    }, prompt_id, user.id);

    return c.json(result);
  } catch (e) {
    console.error(`AI analysis error: ${e.message}`);
    return c.json({ detail: e.message }, 500);
  }
});

// GET /api/ai/history?account_id=X&fund_code=Y&limit=10
ai.get('/history', async (c) => {
  const user = await getCurrentUser(c);
  if (!user) return c.json({ detail: '未登录' }, 401);

  const db = c.env.DB;
  const accountId = c.req.query('account_id');
  const fundCode = c.req.query('fund_code');
  const limit = parseInt(c.req.query('limit') || '10');

  let query = 'SELECT * FROM ai_analysis_history WHERE user_id = ?';
  const params = [user.id];

  if (accountId) {
    query += ' AND account_id = ?';
    params.push(parseInt(accountId));
  }

  if (fundCode) {
    query += ' AND fund_code = ?';
    params.push(fundCode);
  }

  query += ' ORDER BY created_at DESC LIMIT ?';
  params.push(limit);

  const result = await db.prepare(query).bind(...params).all();
  return c.json(result.results);
});

// DELETE /api/ai/history/:id
ai.delete('/history/:id', async (c) => {
  const user = await getCurrentUser(c);
  if (!user) return c.json({ detail: '未登录' }, 401);

  const historyId = parseInt(c.req.param('id'));
  await c.env.DB.prepare(
    'DELETE FROM ai_analysis_history WHERE id = ? AND user_id = ?'
  ).bind(historyId, user.id).run();

  return c.json({ message: '已删除' });
});

// ============================================================================
// Prompt Management
// ============================================================================

// GET /api/ai/prompts
ai.get('/prompts', async (c) => {
  const user = await getCurrentUser(c);
  if (!user) return c.json({ detail: '未登录' }, 401);

  const result = await c.env.DB.prepare(
    'SELECT id, name, system_prompt, user_prompt, is_default, created_at, updated_at FROM ai_prompts WHERE user_id = ? ORDER BY is_default DESC, id'
  ).bind(user.id).all();

  return c.json(result.results);
});

// POST /api/ai/prompts
ai.post('/prompts', async (c) => {
  const user = await getCurrentUser(c);
  if (!user) return c.json({ detail: '未登录' }, 401);

  const { name, system_prompt, user_prompt, is_default } = await c.req.json();
  if (!name) return c.json({ detail: '提示词名称不能为空' }, 400);

  const db = c.env.DB;

  // If setting as default, unset existing defaults
  if (is_default) {
    await db.prepare(
      'UPDATE ai_prompts SET is_default = 0 WHERE user_id = ?'
    ).bind(user.id).run();
  }

  const result = await db.prepare(
    'INSERT INTO ai_prompts (name, prompt, system_prompt, user_prompt, is_default, user_id) VALUES (?, ?, ?, ?, ?, ?)'
  ).bind(name, '', system_prompt || '', user_prompt || '', is_default ? 1 : 0, user.id).run();

  return c.json({ id: result.meta.last_row_id, message: '提示词已创建' });
});

// PUT /api/ai/prompts/:id
ai.put('/prompts/:id', async (c) => {
  const user = await getCurrentUser(c);
  if (!user) return c.json({ detail: '未登录' }, 401);

  const promptId = parseInt(c.req.param('id'));
  const { name, system_prompt, user_prompt, is_default } = await c.req.json();

  const db = c.env.DB;

  if (is_default) {
    await db.prepare(
      'UPDATE ai_prompts SET is_default = 0 WHERE user_id = ?'
    ).bind(user.id).run();
  }

  await db.prepare(
    'UPDATE ai_prompts SET name = ?, system_prompt = ?, user_prompt = ?, is_default = ?, updated_at = datetime() WHERE id = ? AND user_id = ?'
  ).bind(name, system_prompt || '', user_prompt || '', is_default ? 1 : 0, promptId, user.id).run();

  return c.json({ message: '提示词已更新' });
});

// DELETE /api/ai/prompts/:id
ai.delete('/prompts/:id', async (c) => {
  const user = await getCurrentUser(c);
  if (!user) return c.json({ detail: '未登录' }, 401);

  const promptId = parseInt(c.req.param('id'));
  await c.env.DB.prepare(
    'DELETE FROM ai_prompts WHERE id = ? AND user_id = ?'
  ).bind(promptId, user.id).run();

  return c.json({ message: '提示词已删除' });
});

export default ai;
